### DISCUSSION TITLE
Matching Versions (Plugin Development)

### DISCUSSION THREAD

**Discussion by stahlbauer** [swc-project/swc#7166] [Q&A] [answered]
Hi. I am developing a plugin with version "0.69.30" of SWC core. Which version of the NodeJS package is supposed to work with it?

**Comment by stahlbauer** [comment]
To be more concrete: I am facing the following runtime error:

```
Caused by:
    0: Failed to create plugin instance
    1: Error while importing "__wbindgen_placeholder__"."__wbindgen_describe": unknown import. Expected Function(FunctionType { params: [I32], results: [] })', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/swc-0.255.18/src/plugin.rs:228:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
while everything seems to build without any errors.

**Reply by stahlbauer** [reply]
@kdy1 : So was this problem actually related to the issue with `rkyv`? With all the latest dependencies, I am still facing that problem. For some reason, the wasm that was built in my pipeline cannot be invoked by SWC.

**Reply by kdy1** [reply]
Yes, but it's fixed in `rkyv@v0.7.40` and I already verified that it's working locally using `@swc/plugin-jest`.
You should update `@swc/core`, too

**Reply by stahlbauer** [reply]
Is there any point in using `wasm-pack`?

**Reply by kdy1** [reply]
Also, `"__wbindgen_placeholder__"."__wbindgen_describe"` is not related to swc at all. I think it's a bug related to the wasm-pack, but I'm not sure.

**Reply by stahlbauer** [reply]
@kwonoj What is the reason you are using `wasm-pack` for [swc-plugin-coverage-instrument](https://github.com/kwonoj/swc-plugin-coverage-instrument)?

**Reply by stahlbauer** [reply]
(this would also be interesting for the docs on writing plugins)

**Reply by kwonoj** [reply]
I don't use wasm-pack for the coverage plugin except test only fixture setup.

**Reply by stahlbauer** [reply]
Ok, thanks. Was wondering because the WASM created from my plugin has `wbindgen` placeholder functions. See, for example, 
```Imports:
  Functions:
    "env"."__emit_diagnostics": [I32, I32] -> []
    "env"."__set_transform_result": [I32, I32] -> []
    "env"."__set_transform_plugin_core_pkg_diagnostics": [I32, I32] -> []
    "__wbindgen_placeholder__"."__wbindgen_is_undefined": [I32] -> [I32]
    "__wbindgen_placeholder__"."__wbindgen_in": [I32, I32] -> [I32]
    "__wbindgen_placeholder__"."__wbindgen_boolean_get": [I32] -> [I32]
```

**Comment by timofei-iatsenko** [comment]
https://swc.rs/docs/plugin/selecting-swc-core

**Reply by stahlbauer** [reply]
Thanks for the pointer; does not fix my problem, so far. I have tried different version combis. 

For the plugin (`Cargo.toml`):
```
swc = { version = "0.255.3" }
swc_core = {version = "0.69.30", features = [
```
For using the plugin (`package.json`):
```
    "@swc/core": "1.3.40",
```

**Comment by stahlbauer** [comment]
The plugin is called as follows:
```
	return swc.transformSync(sourceContent, {
		filename: sourceFilename,
		sourceMaps: true,
		inlineSourcesContent: false,
		minify: false,
		isModule: true,
		jsc: {
			experimental: {
				plugins: [['TheNameOfMyPlugin', pluginOptions]]
			}
		}
	});
```
By removing the plugin, the error vanishes. But this is not the goal of this exercise :-D

**Comment by stahlbauer** [comment]
The signature of my plugin looks as follows:
```
#[plugin_transform]
pub fn process_transform(
    mut program: Program,
    metadata: TransformPluginProgramMetadata,
) -> Program {
```

The plugin is built using `cargo build --target wasm32-wasi`.

**Comment by timofei-iatsenko** [comment, marked as answer]
it seems there an issue in latest swc. plugins simply doesn't work, and swc authors working on that. You can follow discussing here https://github.com/vercel/next.js/issues/46989#issuecomment-1486989081 and here https://github.com/lingui/swc-plugin/issues/39 and here https://github.com/swc-project/swc/issues/6807#issuecomment-1488195952

**Reply by kdy1** [reply]
The plugins works with `@swc/core`. The problem is `next-swc` which uses recent nightly, which triggers a bug of `rkyv`

**Reply by kdy1** [reply]
I'm working to drop `rkyv` completely, though

**Reply by kdy1** [reply]
I tried various libraries, and all of them had problems so I couldn't use them.
Instead, I updated `rkyv` to `v0.7.40` and it seems like it's working.

**Reply by timofei-iatsenko** [reply]
I saw you mentioned in other discussion that updating to v0.7.40 didn't help, and you tried to implement `rmp_serde`. So it turned out that update helped?

**Reply by kdy1** [reply]
The problem was miscompialtion of nightly rustc, not rkyv@0.7.40.
I think it solves the issue, but I can't be sure at the moment.

**Reply by timofei-iatsenko** [reply]
that's bad because it explodes segmentation of the plugin ecosystem. Right now to support different version of nextjs i need to compile and distribute my plugin with diffrent version of swc_core and rustc.  (and it still will not work for example in plain `@swc/core` because next-swc build with different version of nighly rust)

**Reply by kdy1** [reply]
So what can I do? There's no solution to this problem at the moment.

**Reply by timofei-iatsenko** [reply]
Agree, just adding a bit of input. Hope you find a solution. Great work anyway.
